### PR TITLE
Add Memory-ish protocol

### PR DESCRIPTION
**Pull request by jquense**
Admittedly this is more of a SessionProtocol, It’s probably easy to write a pure memory one in terms of the this, but meh

### CODE REVIEW

**Comment by jquense** [comment]
context, we use this in an Electron app where there is no url bar, the session storage is so ctrl-r works

**Comment by taion** [comment]
I updated the code a bit. How does this look?

File: src/MemoryProtocol.js [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+  _loadState() {
+    try {
+      const { stack, index } = JSON.parse(
+        window.sessionStorage.getItem(STATE_KEY),
+      );
+
+      // Checking instanceof Array is okay because we make the array.
+      if (
+        stack instanceof Array &&
```
Comment (jquense) [inline src/MemoryProtocol.js]: This is still weird :P what's wrong with isArray?

File: src/MemoryProtocol.js [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+    return null;
+  }
+
+  init(delta = 0) {
+    return {
+      ...this._stack[this._index],
+      action: 'POP',
+      index: this._index,
+      delta,
```
Comment (jquense) [inline src/MemoryProtocol.js]: So I originally did this sticking index and delta on location, but what's the value in doing so?

File: src/MemoryProtocol.js (hunk 1 above)
Comment (taion) [reply src/MemoryProtocol.js]: requires a polyfill for older browsers, no? i don't have babel-runtime here... not sure what the right thing is

File: src/MemoryProtocol.js (hunk 2 above)
Comment (taion) [reply src/MemoryProtocol.js]: in principle for consistency with browser protocol... e.g. if you want to animate differently based on whether delta is positive or negative

not sure if index has any purpose... really just from https://github.com/ReactTraining/history/issues/36 and https://github.com/ReactTraining/history/issues/334

File: src/MemoryProtocol.js (hunk 1 above)
Comment (jquense) [reply src/MemoryProtocol.js]: how far back are you going? `isArray` is ie9+

**Comment by taion** [comment]
Good to go on my end

**Comment by jquense** [comment]
lgtm :shipit:

**Pull request merged**